### PR TITLE
Disable typos on exact word

### PR DESCRIPTION
**Pull request by MarinPostma**
This PR introduces the `exact_word` setting to disable typo tolerance on custom words.

If a user query contains a word from `exact_words`, no typo derivation will be made for that particular word.

I have chosen to store the words in a FST, to save on deserialization, and allow for fast lookups.

I had some trouble with the `serde` module, and had to rename it `serde_impl`.

## steps:
- [x] introduce new settings to register words to disable typos on
- [x] in `typos`, return exact match is the current word is part of the word to disable typos for.
- [x] update `Context` to return the exact words dictionary.
- [x] merge #473

### CODE REVIEW

File: milli/src/search/query_tree.rs [hunk 1]
```diff
@@ -333,7 +340,9 @@ fn create_query_tree(
                     children.push(child);
                 }
                 let (word_len_one_typo, word_len_two_typo) = ctx.min_word_len_for_typo()?;
-                let config = TypoConfig { max_typos: 2, word_len_one_typo, word_len_two_typo };
+                let exact_words = ctx.exact_words()?;
+                let config =
+                    TypoConfig { max_typos: 2, word_len_one_typo, word_len_two_typo, exact_words };
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:345]:
```suggestion
                let config =
                    TypoConfig { max_typos: 2, word_len_one_typo, word_len_two_typo, exact_words: ctx.exact_words()? };
```

File: milli/src/search/query_tree.rs [hunk 2]
```diff
@@ -385,8 +394,13 @@ fn create_query_tree(
...
                                 ctx.min_word_len_for_typo()?;
-                            let config =
-                                TypoConfig { max_typos: 1, word_len_one_typo, word_len_two_typo };
+                            let exact_words = ctx.exact_words()?;
+                            let config = TypoConfig {
+                                max_typos: 1,
+                                word_len_one_typo,
+                                word_len_two_typo,
+                                exact_words,
+                            };
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:403]:
```suggestion
                            let config = TypoConfig {
                                max_typos: 1,
                                word_len_one_typo,
                                word_len_two_typo,
                                exact_words: ctx.exact_words()?,
                            };
```

File: milli/tests/search/typo_tolerance.rs [hunk 3]
```diff
@@ -93,3 +98,75 @@ fn test_typo_tolerance_two_typo() {
...
+        let result = search.execute().unwrap();
+        assert_eq!(result.documents_ids.len(), 2);
+    }
+
+    let mut txn = index.write_txn().unwrap();
+
+    let config = IndexerConfig::default();
+    let mut builder = Settings::new(&mut txn, &index, &config);
+    let mut exact_words = BTreeSet::new();
+    // sealand doesn't allow typos anymore
```
Comment (Kerollmops) [inline milli/tests/search/typo_tolerance.rs:159]:
```suggestion
    // zearand doesn't allow typos anymore
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: milli/src/search/query_tree.rs (hunk 1 above)
Comment (MarinPostma) [reply milli/src/search/query_tree.rs:345]: fmt doesn't allow that

File: milli/src/search/query_tree.rs (hunk 2 above)
Comment (MarinPostma) [reply milli/src/search/query_tree.rs:403]: fmt doesn't allow that

File: milli/tests/search/typo_tolerance.rs (hunk 3 above)
Comment (MarinPostma) [inline milli/tests/search/typo_tolerance.rs:159]:
```suggestion
    // `zealand` doesn't allow typos anymore
```

**Review by Kerollmops — APPROVED** [review]
Thank you very much! Ready to be merged :)

**Comment by MarinPostma** [comment]
bors merge

File: milli/src/index.rs [hunk 4]
```diff
@@ -927,6 +928,27 @@ impl Index {
...
         Ok(())
     }
+
+    /// List the words on which typo are not allowed
+    pub fn exact_words<'t>(&self, txn: &'t RoTxn) -> Result<fst::Set<Cow<'t, [u8]>>> {
+        match self.main.get::<_, Str, ByteSlice>(txn, main_key::EXACT_WORDS)? {
+            Some(bytes) => Ok(fst::Set::new(bytes)?.map_data(Cow::Borrowed)?),
+            None => Ok(fst::Set::default().map_data(Cow::Owned)?),
+        }
+    }
```
Comment (Kerollmops) [inline milli/src/index.rs:938]: We MUST return an optional `fst::Set` from this function and never call this default method when executing queries. We should investigate the fact that it is or not the reason for the performance issue!

**Pull request merged**